sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/DeltaMetricStorage.java

```diff
@@ -0,0 +1,138 @@
+
+/**
+ * Allows synchronous collection of metrics and reports delta values isolated by collection handle.
+ *
+ * <p>This storage should allow allocation of new aggregation cells for metrics and unique reporting
+ * of delta accumulations per-collection-handle.
+ */
+@ThreadSafe
+class DeltaMetricStorage<T> {
+  private final Aggregator<T> aggregator;
+  private final ConcurrentHashMap<Attributes, AggregatorHandle<T>> activeCollectionStorage =
+      new ConcurrentHashMap<>();
+  private final List<DeltaAccumulation<T>> unreportedDeltas = new ArrayList<>();
+
+  DeltaMetricStorage(Aggregator<T> aggregator) {
+    this.aggregator = aggregator;
+  }
+
+  /**
+   * Allocates memory for a new metric stream, and returns a handle for synchronous recordings.
+   *
+   * @param attributes The identifying attributes for the metric stream.
+   * @return A handle that will (efficiently) record synchronous measurements.
+   */
+  public BoundStorageHandle bind(Attributes attributes) {
+    AggregatorHandle<T> aggregatorHandle = activeCollectionStorage.get(attributes);
+    if (aggregatorHandle != null && aggregatorHandle.acquire()) {
+      // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+      return aggregatorHandle;
+    }
+
+    // Missing entry or no longer mapped, try to add a new entry.
+    aggregatorHandle = aggregator.createHandle();
+    while (true) {
+      AggregatorHandle<?> boundAggregatorHandle =
+          activeCollectionStorage.putIfAbsent(attributes, aggregatorHandle);
+      if (boundAggregatorHandle != null) {
+        if (boundAggregatorHandle.acquire()) {
+          // At this moment it is guaranteed that the Bound is in the map and will not be removed.
+          return boundAggregatorHandle;
+        }
+        // Try to remove the boundAggregator. This will race with the collect method, but only one
+        // will succeed.
+        activeCollectionStorage.remove(attributes, boundAggregatorHandle);
+        continue;
+      }
+      return aggregatorHandle;
+    }
+  }
+
+  /**
+   * Returns the latest delta accumulation for a specific collection handle.
+   *
+   * @param collector The current reader of metrics.
+   * @param collectors All possible readers of metrics.
+   * @return The delta accumulation of metrics since the last read of a the specified reader.
+   */
+  public synchronized Map<Attributes, T> collectFor(
+      CollectionHandle collector, Set<CollectionHandle> collectors) {
+    // First we force a collection
+    collectSynchronousDeltaAccumulationAndReset();
+    // Now build a delta result.
+    Map<Attributes, T> result = new HashMap<>();
+    for (DeltaAccumulation<T> point : unreportedDeltas) {
+      if (!point.wasReadBy(collector)) {
+        mergeInPlace(result, point.read(collector), aggregator);
+      }
+    }
+    // Now run a quick cleanup of deltas before returning.
+    unreportedDeltas.removeIf(delta -> delta.wasReadyByAll(collectors));
```

Comment:
IIUC, one of the most significant changes from our current model is that rather than clearing all data every collection, we mark each time data has been read, and only after it's been read by all collectors do we clean up. Just want to confirm my understanding

Reply:
Yes.   We store *semi-aggregated* data every time any reader wants to collect metrics, then we reconstitute the full-delta per-reader and join w/ any cumulative storage before reporting.    There's a design "markdown" from Reiley that kind of clarifies this, and I can take a crack at something.  Hoping the new naming helps make this more clear.

Effectively the goal of this class is to get the Synchronous collection path (that's highly performance sensitive) into the reader thread as fast as possible.  We don't care too much about reader contention because we don't expect significant numbers of reader threads.

A few improvements over time:

- I'd like to pre-allocate delta storage and limit its size.  That should be hidden in this class and would be per-metric.
- I'd like to allow multiple reader "within interval" to avoid re-collecting synchronous metrics.
- I'd like to join multiple `MetricExporter`s into a single `MetricReader` if they all use periodic metric reader.

I can open tickets around these improvements, but wanted to make sure we're ok with how this code works now.